src/test/ui/const-generics/mismatched-const-errors.rs

```diff
@@ -0,0 +1,40 @@
+#![feature(generic_const_exprs)]
+#![allow(incomplete_features)]
+
+pub trait Trait {}
+struct HasCastInTraitImpl<const N: usize, const M: usize>;
+impl<const M: usize> Trait for HasCastInTraitImpl<M, { M + 1 }> {}
+pub struct HasTrait<T: Trait>(T);
+
+fn foo1<const N: usize>() -> HasTrait<HasCastInTraitImpl<{ N + 2}, { N }>> {
+    //~^ ERROR mismatched types
+    //~| ERROR unconstrained generic constant
+    loop {}
+}
+
+fn foo2<const N: usize>() -> HasTrait<HasCastInTraitImpl<{ N + 1}, { N + 1 }>> {
+    //~^ ERROR mismatched types
+    //~| ERROR unconstrained generic constant
+    loop {}
+}
+
+fn foo3<const N: usize>() -> HasTrait<HasCastInTraitImpl<{ N + 1}, { N - 1}>> {
+    //~^ ERROR mismatched types
+    //~| ERROR unconstrained generic constant
+    loop {}
+}
+
+// kind of weird that this doesn't compile, although suggestion does compile
+fn foo4<const N: usize>(c : [usize; N]) -> HasTrait<HasCastInTraitImpl<{ N - 1}, { N }>> {
```

Comment:
Why is it that `fn foo4<const N: usize>(c : [usize; N]) -> HasTrait<HasCastInTraitImpl<{ N - 1}, { N - 1 + 1}>> {` does compile, but using `<{ N - 1 }, { N }>` fails?

Reply:
generic_const_exprs does not really do any kind of "smart" equating of consts. we pretty much just directly check that the `AbstractConst` trees are exactly the same. `N - 1 + 1` and `N` have different trees so are considered unequal. (Honestly they might not even be equal anyway if we have panic on overflow, i.e. if `N` was `0`)